harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -118,9 +190,91 @@ def _set_random_seeds(cls: Type["PyTorchTrialController"], seed: int) -> None:
         # torch.backends.cudnn.deterministic = True
         # torch.backends.cudnn.benchmark = False
 
+    def report_training_metrics(self, steps_completed, metrics):
+        metrics = self._prepare_metrics(metrics)
+        self.metric_writer.on_train_step_end(
+            self.steps_completed,
+            metrics["avg_metrics"],
+            metrics["batch_metrics"],
+        )
+
+        # Only report on the chief worker
+        if self.is_chief:
+            avg_metrics = metrics.get("avg_metrics", {})
+            batch_metrics = metrics.get("batch_metrics", [])
+            self.core_context.train.report_training_metrics(
+                steps_completed=steps_completed, metrics=avg_metrics, batch_metrics=batch_metrics
+            )
+
+    def report_validation_metrics(self, steps_completed, metrics):
+        if self.is_chief:
+            self.core_context.train.report_validation_metrics(steps_completed, metrics)
+
+    def on_epoch_start(self, epoch_idx: int):
```

Comment:
Is there a criteria for what things to on the TrainLoop and what things live on the TrialController?

Right now the abstraction feels really leaky.

Reply:
TrainLoop merged into TrialController. the abstraction isn't great, but i basically wanted to separate out the training loop variables/constants from the outer initialization layer.